starter-code/TicTacToe/app/src/main/java/ly/generalassemb/drewmahrt/tictactoe/GameActivity.java

```diff
@@ -1,13 +1,148 @@
+        mSlot2_2.setOnClickListener(gridListener(mTicTacToe.getCoordinate(8)));
+    }
+
+    public void changeIcon(Coordinate coordinate,ImageView imageView, int turn) {
+        if (coordinate.isEmpty()) {
+
+            if ( !(mTicTacToe.isWinner(mTicTacToe.getPlayer1())
+                    || mTicTacToe.isWinner(mTicTacToe.getPlayer2()) )
+                    ) {
+                if (turn % 2 == 1) {
+                    imageView.setImageResource(R.drawable.ic_o);
+                    coordinate.setStatus(1);
+                    mTicTacToe.getPlayer1().addItem(coordinate);
+                    mTurnText.setText(mTicTacToe.getPlayer2().getName() + "'s Turn.");
+                    mTurn++;
+                } else {
+                    imageView.setImageResource(R.drawable.ic_x);
+                    coordinate.setStatus(2);
+                    mTicTacToe.getPlayer2().addItem(coordinate);
+                    mTurnText.setText(mTicTacToe.getPlayer1().getName() + "'s Turn.");
+                    mTurn++;
+                }
+            }
+        }
+    }
+
+    //Use changeIcon for the listener
+    public View.OnClickListener gridListener (final Coordinate coordinate){
+        View.OnClickListener onClickListener = new View.OnClickListener() {
+            @Override
+            public void onClick(View v) {
+                switch (v.getId()){
+                    case R.id.grid_0_0:
+                        changeIcon(coordinate,(ImageView)v, mTurn);
+                        break;
+                    case R.id.grid_0_1:
+                        changeIcon(coordinate,(ImageView)v, mTurn);
+                        break;
+                    case R.id.grid_0_2:
+                        changeIcon(coordinate,(ImageView)v, mTurn);
+                        break;
+                    case R.id.grid_1_0:
+                        changeIcon(coordinate,(ImageView)v, mTurn);
+                        break;
+                    case R.id.grid_1_1:
+                        changeIcon(coordinate,(ImageView)v, mTurn);
+                        break;
+                    case R.id.grid_1_2:
+                        changeIcon(coordinate,(ImageView)v, mTurn);
+                        break;
+                    case R.id.grid_2_0:
+                        changeIcon(coordinate,(ImageView)v, mTurn);
+                        break;
+                    case R.id.grid_2_1:
+                        changeIcon(coordinate,(ImageView)v, mTurn);
+                        break;
+                    case R.id.grid_2_2:
+                        changeIcon(coordinate,(ImageView)v, mTurn);
+                        break;
+                }
+                if (!mIsFinished) {
+                    mIsFinished = checkWinner();
+                }
+            }
+        };
+        return onClickListener;
+    }
+
+    public boolean checkWinner(){
+    //Check for win condition or game is done
+        if (mTicTacToe.isWinner(mTicTacToe.getPlayer1())) {
```

Comment:
Interesting approach to move some of the game logic to your singleton class - it makes this activity clean and easy to read!